File: v170/cgi-bin/webnotes/modules/System/Sales and Purchase Return Wizard.js
cur_frm.cscript.onload = function(doc,dt,dn){
  if(!doc.return_date) set_multiple(dt,dn,{return_date:get_today()});
  doc.delivery_note_no = '';
  doc.purchase_receipt_no = '';
  doc.return_type ='';
}

cur_frm.cscript.return_type = function(doc, cdt, cdn) {
  if(doc.return_type == 'Sales Return'){
    unhide_field('delivery_note_no','Delivery Note No');
    hide_field('purchase_receipt_no','Purchase Receipt No');
    unhide_field('Make Credit Note');
    hide_field('Make Debit Note');
  } 
  if(doc.return_type == 'Purchase Return'){
    unhide_field('purchase_receipt_no','Purchase Receipt No');
    hide_field('delivery_note_no','Delivery Note No');
    unhide_field('Make Debit Note');
    hide_field('Make Credit Note');
  }
  unhide_field('Make JV');
  unhide_field('Get Items');  
  unhide_field('return_details','Return Details');
  unhide_field('Make Stock Entry');

}

cur_frm.cscript['Get Items'] = function(doc, cdt, cdn) {
  var callback = function(r,rt) {
    cur_frm.cscript.max_returned = r.message;
    cur_frm.cscript.dno(doc,cdt,cdn);
  }
  $c_obj(make_doclist(doc.doctype, doc.name),'pull_details','',callback);

}

cur_frm.cscript.dno = function(doc,cdt,cdn){
  if(doc.delivery_note_no){
    cur_frm.cscript.doc_no = doc.delivery_note_no;
  }
  else if(doc.purchase_receipt_no){
    cur_frm.cscript.doc_no = doc.purchase_receipt_no;
  }
  doc.delivery_note_no = ''
  doc.purchase_receipt_no = '' 
  refresh_field('delivery_note_no');
  refresh_field('purchase_receipt_no');
  refresh_field('return_details');
}
cur_frm.cscript['Make Stock Entry'] = function(doc, dt, dn) {

  cur_frm.cscript.make_se(doc,dt,dn);
}

cur_frm.cscript.make_se = function(doc, dt, dn){


  var se = LocalDB.create('Stock Entry');
  se = locals['Stock Entry'][se];

  var comp_name = cur_frm.cscript.max_returned.length;
  se.posting_date = dateutil.obj_to_str(new Date());
  se.transfer_date = dateutil.obj_to_str(new Date());
  se.remarks = doc.return_type + ' of ' + cur_frm.cscript.doc_no;

  se.company = cur_frm.cscript.max_returned[comp_name-1];
  se.fiscal_year = sys_defaults.fiscal_year;
  se.purpose = doc.return_type;
  var flg  = 0
  var doc = locals[doc.doctype][doc.name];
  el = getchildren('Return Detail',doc.name,'return_details');
  var len = el.length;
  try{
    for(var i = 0; i< len; i++){
      if(el[i].returned_qty && el[i].returned_qty <= cur_frm.cscript.max_returned[i]){
        var d1 = LocalDB.add_child(se, 'Stock Entry Detail', 'mtn_details');
        d1.detail_name = el[i].detail_name;
        d1.item_code = el[i].item_code;
        d1.description = el[i].description;
        d1.transfer_qty = el[i].returned_qty;
        d1.stock_uom = el[i].uom;
        
      }
      else if(el[i].returned_qty && el[i].returned_qty > cur_frm.cscript.max_returned[i]){
        throw('You can return a maximum of '+ cur_frm.cscript.max_returned[i]+' qty for tem: '+el[i].item_code);
      }
      else{
        throw('Please Enter number of quantities to be returned');
      }
    }
  }
  catch(e){
    alert("Error : "+e);
    flg = 1;
  }
  if(flg == 0){
    loaddoc('Stock Entry', se.name);
  }
  
}

cur_frm.cscript['Make JV'] = function(doc, dt, dn) {
  var excise = LocalDB.create('Journal Voucher');
  excise = locals['Journal Voucher'][excise];
  
  loaddoc('Journal Voucher',excise.name);
}

cur_frm.cscript['Make Debit Note'] = function(doc, dt, dn) {

  var calback = function(r,rt){
    var ent_det = r.message;
    
    var debit = LocalDB.create('Journal Voucher');
    debit = locals['Journal Voucher'][debit];
    
    debit.voucher_type = 'Debit Note';
    debit.company = ent_det[0];
    debit.fiscal_year = sys_defaults.fiscal_year;
    debit.is_opening = 'No';
    debit.posting_date = dateutil.obj_to_str(new Date());
    debit.voucher_date = dateutil.obj_to_str(new Date());
    var d1 = LocalDB.add_child(debit,'Journal Voucher Detail','entries')
    d1.account = ent_det[1]
    loaddoc('Journal Voucher',debit.name);
  }
  $c_obj(make_doclist(doc.doctype, doc.name),'get_voucher_det','',calback);
  
}

cur_frm.cscript['Make Credit Note'] = function(doc, dt, dn) {

  var c_back = function(r,rt){
    var ent_det = r.message;
    
    var credit = LocalDB.create('Journal Voucher');
    credit = locals['Journal Voucher'][credit];

    
    credit.voucher_type = 'Credit Note';
    credit.company = ent_det[0];
    credit.fiscal_year = sys_defaults.fiscal_year;
    credit.is_opening = 'No';
    credit.posting_date = dateutil.obj_to_str(new Date());
    credit.voucher_date = dateutil.obj_to_str(new Date());
    var d1 = LocalDB.add_child(credit,'Journal Voucher Detail','entries')
    d1.account = ent_det[1]
    loaddoc('Journal Voucher',credit.name);
  }  
  $c_obj(make_doclist(doc.doctype, doc.name),'get_voucher_det','',c_back);
}

File: modules/System/Sales and Purchase Return Wizard/Sales and Purchase Return Wizard.py
class DocType :
  def __init__(self, doc, doclist=[]):
    self.doc = doc
    self.doclist = doclist
      
  def validate(self):
    if not self.doc.return_type:
      msgprint("Please Enter Return Type")
      raise Exception
      
  def pull_details(self):
    flg = 0
    ret_type = self.doc.return_type
    if ret_type == 'Sales Return' and self.doc.delivery_note_no:
      det = sql("select name, item_code, description, qty, uom from `tabDelivery Note Packing Detail` where parent = '%s' "%(self.doc.delivery_note_no))
      cust_supp = sql("select company from `tabDelivery Note` where name = '%s'"%(self.doc.delivery_note_no))
      flg = 1
    if ret_type == 'Purchase Return' and self.doc.purchase_receipt_no:
      det = sql("select name, item_code, description, received_qty, uom from `tabPurchase Receipt Detail` where parent = '%s' "%(self.doc.purchase_receipt_no))
      cust_supp = sql("select company from `tabPurchase Receipt` where name = '%s'"%(self.doc.purchase_receipt_no))
      flg = 1
    if ret_type == 'Sales Return' and flg == 0:
      msgprint("Please Enter Delivery Note No")    
      raise Exception
    if ret_type == 'Purchase Return' and flg == 0:
      msgprint("Please Enter Purchase Receipt No")
      raise Exception
      
    if flg == 1:
      ret = self.add_details(det)
      ret.append(cust_supp[0][0])
      self.doc.save()
      return ret

    
    
  def add_details(self, det):
    r = []
    ret = 0
    self.doc.clear_table(self.doclist, 'return_details', 1)
    for i in det:
      prev_ret = sql("select transfer_qty from `tabStock Entry Detail` where detail_name = '%s' and docstatus = 1 "%(i[0]))
      for prev in prev_ret:
        ret += flt(prev and prev[0] or 0)
      ret_child = addchild(self.doc, 'return_details', 'Return Detail', 1, self.doclist)
      
      ret_child.detail_name = i and i[0] or ''
      ret_child.item_code = i and i[1] or ''
      ret_child.description = i and i[2] or ''
      ret_child.qty = i and flt(i[3]) or 0
      ret_child.uom = i and i[4] or ''
      ret_child.save()
      if prev_ret:
        r.append(flt(i[3])-flt(ret))
      else:
        r.append(i and flt(i[3]) or 0)
    self.doc.save()
    return r
    
    
  def get_voucher_det(self):
    fg = 0
    det_list = []
    if self.doc.return_type == 'Sales Return':
      if self.doc.delivery_note_no:
        det = sql("select customer_name, company from `tabDelivery Note` where name = '%s'"%(self.doc.delivery_note_no))
        det_list.append(det[0][1])
        acc_head = sql("select name from `tabAccount` where account_name = '%s'"%(det[0][0]))
        if acc_head:
          det_list.append(acc_head[0][0])
        else:
          msgprint("Account Head of customer "+det[0][0]+" does not exists. Please create account head before making Credit Note")
          raise Exception
        fg = 1
      else:
        msgprint("Please Enter Delivery Note No to make Credit Note")
        raise Exception
    elif self.doc.return_type == 'Purchase Return':
      if self.doc.purchase_receipt_no:
        det = sql("select supplier, company from `tabPurchase Receipt` where name = '%s'"%(self.doc.purchase_receipt_no))
        det_list.append(det[0][1])        
        acc_head = sql("select name from `tabAccount` where account_name = '%s'"%(det[0][0]))
        if acc_head:
          det_list.append(acc_head[0][0])
        else:
          msgprint("Account Head of supplier "+det[0][0]+" does not exists. Please create account head before making Debit Note")
          raise Exception
        fg = 1 
      else:
        msgprint("Please Enter Purchase Receipt No to make debit note")
        raise Exception
    if fg == 1:
      return det_list